### ISSUE TITLE
Safari Unrecognized Content-Security-Policy directive 'report-to' renew token bug.

### ISSUE THREAD

**Issue by WojciechSlyp** [okta/okta-auth-js#257]
Hi!

I've encountered the following bug:
- I get the cached token with `const idToken = await okta.tokenManager.get('id_token')`
- I try to renew it with `okta.token.renew(idToken)`

Safari gives me:
`Unrecognized Content-Security-Policy directive 'report-to'` error and renew promise is rejected with:
`error.errorCode: login_required, error.description: The client specified not to prompt, but the user is not logged in.`

This only happens on Safari (tested on 13.0.2), macos and iOS 13.1.
Chrome, FF etc work fine.

**Comment by swiftone** [comment]
Hi @WojciechSlyp - Our CSP settings are (for now) to report on violations (as opposed to fail), and the directive you list is only (reportedly) supported by Chrome so far ( https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to ), however the unsupported setting should NOT cause fatal errors.

We will look into recreating this, but our initial thought is that the failure to renew and the CSP directive mention is a distraction, not a cause.

**Comment by swiftone** [comment]
internal-ref: 258252

**Comment by Jaffacakes82** [comment]
I'm getting this same error on Safari 12 using the Okta-hosted sign in screen. Can you advise if there's a way to resolve this in Safari?

Thanks.

**Comment by swiftone** [comment]
@Jaffacakes82 - Safari doesn't support the report-to CSP directive (as mentioned in an earlier comment) but those directives should not cause any error.

We suspect that warning is unrelated to the actual problem you are encountering.  Whatever problem that is, that is giving you the `error.errorCode: login_required, error.description: The client specified not to prompt, but the user is not logged in.` error.

Are you seeing that error, or only the warning?  That error message can occur if a session is sufficiently old (how old depends on your specific settings)  Can you describe the events that lead to the message?

**Comment by swiftone** [comment]
Closing as unable to reproduce - feel free to reopen with additional details.

**Comment by advancedvdragutinovic** [comment]
Hello, I am having similar issue in Safari. 
We are implementing Okta SSO using okta-auth-js in our Angular app.

**Comment by advancedvdragutinovic** [comment]
Hello, I am having a similar issue in Safari (Version 13.0.5 (15608.5.11))

We are integrating Okta within our Angular app.

When calling OktaAuth.session.me() to get the current session info, I am seeing the following error in Safari's console: 
`Unrecognized Content-Security-Policy directive 'report-to'.`
After that, there is a 404 response for (https://{myOkta}.com/api/v1/sessions/me). 

Before calling the session.get(), we are calling:
`OktaAuth.signIn()` and `OktaAuth.token.getWithoutPrompt()`, 
which both complete successfully (i.e I can see the access_token and id_token in the response, along with other info). 

Session response: 
```
{
    "errorCode": "E0000007",
    "errorSummary": "Not found: Resource not found: me (Session)",
    "errorLink": "E0000007",
    "errorId": "oaemp3AIUP3RxqefXww2veR0Q",
    "errorCauses": []
}
```

This works fine in both Chrome and Firefox. 
Could this be a Safari related bug? 

Thanks, 
Velibor D

**Comment by swiftone** [comment]
@advancedvdragutinovic - Thanks for the report.  Let's see what we can do to help.

First: For future reference, comments on closed issues can be overlooked - Better to reopen or create a new issue.

Second: As per above, the CSP policy should be unrelated: it's a report, but has no impact on the connection.

Third: If your Safari is set to block 3rd party cookies (which I believe is the default in recent Safari versions) you'll be unable to get a success response from the https://{myOktadomain}.com/api/v1/sessions/me endpoint (in this case, the user's browser is the 1st party, your page domain is the 2nd party, and the the okta domain is the 3rd party.  

As you see from your other calls, this doesn't mean you can't talk to Okta services, it just means you can't talk to services that require cookies.  The majority of our service calls use a bearer authentication token in the header, but the /sessions/me endpoint is a special case (the purpose is to check for the cookie on the browser-side, so it necessarily uses cookies).

What is your use-case that you are calling sessions.me() for?

**Comment by advancedvdragutinovic** [comment]
Hello, and thanks for a prompt response, and advices.  

You were right, my Safari did have 3rd party cookies disabled. After enabling them, it's working as expected. Thank you! 

Our use-case is following: 
We have multiple modules (apps), that the user can navigate to after logging in from our main app. These additional apps won't be communicating with Okta, instead they will continue using our internal tokens for authentication/authorization with our APIs. Our intention is to store the user's sessionId in our APIs (after user logs in), so we can extend the session for them when they are active in any of the modules. Thus, if they come back to the main app, their session would still be active. 

Is there any other way to obtain the sessionId which doesn't require use of cookies? I feel we will have trouble if we rely on using cookies since not everyone will have them enabled. 

Thanks,
Velibor D

**Comment by swiftone** [comment]
@advancedvdragutinovic - Your use case sounds like there may be an alternative to using the sessionID, but that's outside the scope of our SDK maintenance.  I suggest you reach out to our support team at developers@okta.com and describe your need to them - they can delve into their better-than-my storehouse of solutions to better guide you.

Hope that helps, and if there is an SDK issue please open an issue with details!

**Comment by davoam** [comment]
> Hello, I am having a similar issue in Safari (Version 13.0.5 (15608.5.11))
> (okta-auth-js-3.0.1)
> 
> We are integrating Okta within our Angular app.
> 
> When calling OktaAuth.session.me() to get the current session info, I am seeing the following error in Safari's console:
> `Unrecognized Content-Security-Policy directive 'report-to'.`
> After that, there is a 404 response for (https://{myOkta}.com/api/v1/sessions/me).
> 
> Before calling the` session.get()`, we are calling:
> `OktaAuth.signIn()` and `OktaAuth.token.getWithoutPrompt()`,
> which both complete successfully (i.e I can see the access_token and id_token in the response, along with other info).
> 
> Session response:
> 
> ```
> {
>     "errorCode": "E0000007",
>     "errorSummary": "Not found: Resource not found: me (Session)",
>     "errorLink": "E0000007",
>     "errorId": "oaemp3AIUP3RxqefXww2veR0Q",
>     "errorCauses": []
> }
> ```
> 
> This works fine in both Chrome and Firefox.
> Could this be a Safari related bug?
> 
> Thanks,
> Velibor D

We have the same problem

**Comment by StevenCountrymanCFA** [comment]
Bump!

We're experiencing the same issue on Safari as well. 

1. We first check for an active session with the /me endpoint

1. Then we show the signin widget if the response is a 404 (if there is no session or if Safari blocks it). 

1. The user then uses the signin widget to create a new Okta session and then we attempt to get a new token using the AuthJS library with "getWithPopup" and/or "getWithoutPrompt", both of which are being blocked with the same errors reported above: 

- Unrecognized Content-Security-Policy directive 'report-to'. 

- error.errorCode: login_required, error.description: The client specified not to prompt, but the user is not logged in.

**Comment by swiftone** [comment]
@StevenCountrymanCFA - thanks for the report and the good description of your steps.  Let me break it down:


- Looks like you are handling the 404 for the /sessions/me correctly (For anyone else just coming to this, Safari is blocking third party cookies by default now, which means the /sessions/me endpoint is unavailable as it relies on a cookie set by the issuer domain)

- The Content-Security-Policy should not be relevant, the CSP is set to report issues (not enforce), so Safari not understanding that particular directive isn't (currently) meaningful.  

- I think the remaining issue is that you have both the SignIn Widget (which has a copy of okta-auth-js) and okta-auth-js directly.  We've uncovered conflicts between these and recently put out what should be a fix.  If you make sure your direct use of okta-auth-js is using 3.2.2+, that should address the issue.  

Please let us know if this fixes your issue or not.


TL;DR: Update your okta-auth-js to 3.2.2+

**Comment by StevenCountrymanCFA** [comment]
@swiftone Thanks for the quick reply!

I was on 3.2.1 of the okta-auth-js library so I'm updating right now to try that out :D

Is there a specific version of the signin widget I need to be on as well? Or does the new version of the okta-auth-js library fix it no matter what?

**Comment by swiftone** [comment]
> Is there a specific version of the signin widget I need to be on as well?

You should not need to update the SIW to any specific version, this change is meant to address the incompatibility without any change there.

**Comment by StevenCountrymanCFA** [comment]
@swiftone 

Can confirm 3.2.2 does indeed fix the incompatibility with the signin widget :D It can now renew tokens from its own redirect AND tokens saved via the signin widget, which was something that didn't work before.

Thank you!
